File: perl/hamming74.pl
#!/usr/bin/perl -w

use strict;
use Getopt::Long;

my ($json, $tex, $help);
my $usage = "Usage: $0 [options]\n"
    .       "          --tex,-t  Add LaTex transition labels to DOT output\n"
    .       "         --json,-j  Print transducer in JSON format, not GraphViz DOT\n"
    .       "         --help,-h  Print this help text\n"
    ;

GetOptions ("json" => \$json,
	    "tex" => \$tex,
	    "help" => \$help)
    or die $usage;

if ($help) { print $usage; exit }

my (%trans, %stateIndex, @stateName);

header();

for (my $n = 0; $n < 2; ++$n) {
    trans ("S", bin($n,1), $n % 2, undef);
}

for (my $n = 0; $n < 4; ++$n) {
    trans (bin($n>>1,1), bin($n,2), $n % 2, undef);
}

for (my $n = 0; $n < 8; ++$n) {
    trans (bin($n>>1,2), bin($n,3), $n % 2, undef);
}

for (my $n = 0; $n < 16; ++$n) {
    trans (bin($n>>1,3), "p1_".bin($n,4), $n % 2, parity($n,1,2,4));
}

for (my $n = 0; $n < 16; ++$n) {
    trans ("p1_".bin($n,4), "p2_".bin($n,4), undef, parity($n,1,3,4));
}

for (my $n = 0; $n < 16; ++$n) {
    trans ("p2_".bin($n,4), "d1_".bin($n,3), undef, bit($n,1));
}

for (my $n = 0; $n < 8; ++$n) {
    trans ("d1_".bin($n,3), "p3_".bin($n,3), undef, parity($n,2,3,4));
}

for (my $n = 0; $n < 8; ++$n) {
    trans ("p3_".bin($n,3), "d2_".bin($n,2), undef, bit($n,2));
}

for (my $n = 0; $n < 4; ++$n) {
    trans ("d2_".bin($n,2), "d3_".bin($n,1), undef, bit($n,3));
}

for (my $n = 0; $n < 2; ++$n) {
    trans ("d3_".bin($n,1), "S", undef, bit($n,4));
}

state("S","S");

for (my $n = 0; $n < 2; ++$n) {
    state (bin($n,1));
    state ("d3_".bin($n,1));
}

for (my $n = 0; $n < 4; ++$n) {
    state (bin($n,2));
    state ("d2_".bin($n,2));
}

for (my $n = 0; $n < 8; ++$n) {
    state (bin($n,3));
    state ("d1_".bin($n,3));
    state ("p3_".bin($n,3));
}

for (my $n = 0; $n < 16; ++$n) {
    state ("p1_".bin($n,4));
    state ("p2_".bin($n,4));
}

if ($json) {
    state("E");
}

footer();

sub header {
    if (!$json) {
	print "digraph G {\n";
	print "rankdir=LR;\n";
    }
}

sub trans {
    my ($src, $dest, $in, $out) = @_;
    if ($json) {
	push @{$trans{$src}}, [$in,$out,$dest];
    } else {
	if ($tex) {
	    print "$src -> $dest [label=\$", defined($in) ? "${in}_2" : "\\epsilon", "/", defined($out) ? "${out}_2" : "\\epsilon", "\$];\n";
	} else {
	    print "$src -> $dest [label=\"", defined($in) ? "${in}" : "", "/", defined($out) ? "${out}" : "", "\"];\n";
	}
    }
}

sub state {
    my ($state, $label) = @_;
    $label = "" unless defined $label;
    if ($json) {
	$stateIndex{$state} = @stateName;
	push @stateName, $state;
    } else {
	print "$state [label=\"$label\"];\n";
    }
}

sub footer {
    if ($json) {
	print "{\"state\": [\n";
	for my $s (0..$#stateName) {
	    my $id = $stateName[$s];
	    print
		" {\"n\":$s,\"id\":\"$id\",\"trans\":[",
		join (",", map ("{"
				. (defined($$_[0])?"\"in\":\"$$_[0]\",":"")
				. (defined($$_[1])?"\"out\":\"$$_[1]\",":"")
				. "\"to\":" . $stateIndex{$$_[2]}
				. "}",
				@{$trans{$id}})),
		"]}", ($s < $#stateName ? "," : ""), "\n";
	}
	print "]}\n";
    } else {
	print "}\n";
    }
}

sub bin {
    my ($n, $d) = @_;
    my $s = "";
    for (my $k = 0; $k < $d; ++$k) {
	$s = (($n % 2) ? "1" : "0") . $s;
	$n = int ($n / 2);
    }
    return $s;
}

sub bit {
    my ($n, $pos) = @_;
    return ($n & (1 << (4-$pos))) ? 1 : 0;
}

sub parity {
    my ($n, @pos) = @_;
    my $p = 0;
    for my $pos (@pos) { $p += bit($n,$pos) }
    return $p % 2;
}

File: python/hamming74.py
#!/usr/bin/env python3

import argparse
import sys

def main():
    parser = argparse.ArgumentParser(description='Generate Hamming(7,4) WFST')
    parser.add_argument('--json', '-j', action='store_true', help='Print transducer in JSON format, not GraphViz DOT')
    parser.add_argument('--tex', '-t', action='store_true', help='Add LaTeX transition labels to DOT output')
    args = parser.parse_args()

    trans_map = {}
    state_index = {}
    state_names = []

    def binstr(n, d):
        s = ''
        for _ in range(d):
            s = ('1' if n % 2 else '0') + s
            n //= 2
        return s

    def bit(n, pos):
        return 1 if (n & (1 << (4 - pos))) else 0

    def parity(n, *positions):
        return sum(bit(n, p) for p in positions) % 2

    def add_trans(src, dest, inp, out):
        if args.json:
            trans_map.setdefault(src, []).append((inp, out, dest))
        else:
            if args.tex:
                in_label = f'{inp}_2' if inp is not None else '\\epsilon'
                out_label = f'{out}_2' if out is not None else '\\epsilon'
                print(f'{src} -> {dest} [label=${in_label}/{out_label}$];')
            else:
                in_label = str(inp) if inp is not None else ''
                out_label = str(out) if out is not None else ''
                print(f'{src} -> {dest} [label="{in_label}/{out_label}"];')

    def add_state(state, label=None):
        if label is None:
            label = ''
        if args.json:
            state_index[state] = len(state_names)
            state_names.append(state)
        else:
            print(f'{state} [label="{label}"];')

    # Header
    if not args.json:
        print('digraph G {')
        print('rankdir=LR;')

    # Build transitions
    for n in range(2):
        add_trans('S', binstr(n, 1), n % 2, None)

    for n in range(4):
        add_trans(binstr(n >> 1, 1), binstr(n, 2), n % 2, None)

    for n in range(8):
        add_trans(binstr(n >> 1, 2), binstr(n, 3), n % 2, None)

    for n in range(16):
        add_trans(binstr(n >> 1, 3), 'p1_' + binstr(n, 4), n % 2, parity(n, 1, 2, 4))

    for n in range(16):
        add_trans('p1_' + binstr(n, 4), 'p2_' + binstr(n, 4), None, parity(n, 1, 3, 4))

    for n in range(16):
        add_trans('p2_' + binstr(n, 4), 'd1_' + binstr(n, 3), None, bit(n, 1))

    for n in range(8):
        add_trans('d1_' + binstr(n, 3), 'p3_' + binstr(n, 3), None, parity(n, 2, 3, 4))

    for n in range(8):
        add_trans('p3_' + binstr(n, 3), 'd2_' + binstr(n, 2), None, bit(n, 2))

    for n in range(4):
        add_trans('d2_' + binstr(n, 2), 'd3_' + binstr(n, 1), None, bit(n, 3))

    for n in range(2):
        add_trans('d3_' + binstr(n, 1), 'S', None, bit(n, 4))

    # Register states
    add_state('S', 'S')

    for n in range(2):
        add_state(binstr(n, 1))
        add_state('d3_' + binstr(n, 1))

    for n in range(4):
        add_state(binstr(n, 2))
        add_state('d2_' + binstr(n, 2))

    for n in range(8):
        add_state(binstr(n, 3))
        add_state('d1_' + binstr(n, 3))
        add_state('p3_' + binstr(n, 3))

    for n in range(16):
        add_state('p1_' + binstr(n, 4))
        add_state('p2_' + binstr(n, 4))

    if args.json:
        add_state('E')

    # Footer
    if args.json:
        print('{"state": [')
        for s in range(len(state_names)):
            sid = state_names[s]
            trans_list = trans_map.get(sid, [])
            parts = []
            for inp, out, dest in trans_list:
                t = '{'
                fields = []
                if inp is not None:
                    fields.append(f'"in":"{inp}"')
                if out is not None:
                    fields.append(f'"out":"{out}"')
                fields.append(f'"to":{state_index[dest]}')
                t += ','.join(fields) + '}'
                parts.append(t)
            comma = ',' if s < len(state_names) - 1 else ''
            print(f' {{"n":{s},"id":"{sid}","trans":[{",".join(parts)}]}}{comma}')
        print(']}')
    else:
        print('}')


if __name__ == '__main__':
    main()
